Extracted MGL Sections

nesanders opened the pull request:
Migrated from https://github.com/vynpt/ml-maple-bill-summarization/pull/15

> extract_MGL_sections.ipynb --> Added code to extract sections/chapter from the bills
App3.py --> Implemented only using 12 bills, pulls texts from the sections of a bill selected rather than all the bills (as in app2.py) for the prompt
App4.py --> Includes all the bills --> pulls text from the sections of a bill selected (~1300 bills don't have any sections data); includes code to pull the parquet file from my google drive for now as it is too large for GitHub.
MGL_analysis.ipynb --> Started a python notebook to evaluate missing sections, words length in the extracted sections, etc.

mmomin15 committed:
Add code for extracting sections

mmomin15 committed:
App3 - contains only 12 bills with more sections than App2; also included MGL in the prompt to extract categories

App3 - contains all sections (1300 bill don't have MGL Sections data)

mmomin15 committed:
updated code to use parque version of the MGL sections file.

mmomin15 committed:
updated code to use parque version of the MGL sections file instead of CSV.

nesanders commented on demoapp/app3.py:
We ultimately want to develop and maintain a single version of the demo app. If v1, v2, v3 are strictly vestigial, then I recommend just deleting them as part of this PR and renaming v4 to something more generic (like `demo_app.py`).

mmomin15 replied:
Makes sense. Will do that.

nesanders commented on demoapp/app4.py:
Assuming you started from `app2.py`, I would ask that you replace that file with these changes rather than create a new file. That will make your line-level changes show up in the PR diff and make it much more feasible to review and offer feedback. (Right now, I am not sure what changes this version implements,)

Please also add top level multi-line comment to this file explaining the purpose of the app and any major features you've implemented.

nesanders commented on demoapp/MGL_Analysis.ipynb:
Please add toplevel comment to this notebook explaining its purpose and whether it is of ongoing use or just a one-off exploration

mmomin15 replied:
Ok, added comment at the top of the notebook.

nesanders commented on demoapp/extract_MGL_sections.ipynb:
IIUC this notebook implements the critical functionality of extracting section references from bills and pulling relevant section text!  This is super important.

I recommend migrating this functionality to a plain python script (`.py`). There are a few reasons not to use notebooks for this.

1. Python scripts can be used more programmatically, so that we can more easily rerun these functions and integrate them with our codebase in the future.
1. GitHub does not support inline commenting and diffs in the notebooks in the PR review process.

From what I see in the notebook, I have a few general comments. When migrated to a script, I can more easily share line-specific feedback.

General comments:

1. Please add docstrings to functions explaining their functions and documenting any non-obvious arguments.
2. I recommend adding python [typehints](https://dagster.io/blog/python-type-hinting) (input and output) to all function signatures.
3. The loop beginning with `for idx, doc in empty_lists["DocumentText"].items():` looks like it is basically regularizing the extracted chapter/section references. This could be wrapped into a function!

mmomin15 replied:
This is done. Moved all the code for extracting and processing the MGLs into an extract_mgl_sections.py file. Added docstrings to each function. Also modified the extract_MGL_sections.ipynb notebook to just include the example code for using the functions to extract MGL sections.

nesanders replied:
Nice, thank you!

nesanders reviewed (commented):
This is great, thanks, @mmomin15 ! Super important functionality for us and I am excited to make it broadly avaialble!

There are a few little formatting aspects of the PR that make it difficult for me to offer detailed feedback, so I am leaving some comments to request a bit of reformatting that would allow me to comment on the details better.

mmomin15 committed:
Moved all the code to extract the MGL sections, combine MGL text from a list of strings into one string, and MGL word count to extract_mgl_sections.py for better integration

extract_MGL_sections.ipynb now has an example code to run the function to extract MGL chapter-sections and get MGL text from the API

Deteled app3.py (originally copied from app2.py) and added changes made to app3.py to app2.py

Renamed app4.py to demo_app.py

Deleted MGL_Analysis.ipynb for now; will add it back with more analysis in the future; updated wordcount

Updated columns names in 12_bills_with_mgl_with_more_sections.csv for consistency. Added a column where no MGL was found through API

mmomin15 committed:
Added text explaining the goal of the notebook

mmomin15 committed:
Added a note at the top explaining the goal of this notebook

mmomin15 committed:
Added a couple of sentences explaining the goal and changes made to these files.

mmomin15 committed:
added more detailed explanation

nesanders commented on demoapp/app2.py:
```diff
@@ -15,10 +15,16 @@
 from langchain.embeddings.openai import OpenAIEmbeddings
 from langchain.text_splitter import CharacterTextSplitter
 from langchain.document_loaders import TextLoader
+from langchain.docstore.document import Document
```
Is it iontentional that your PR introduces changes to `app2.py`? Or should all your changes go in `demo_app.py`?

nesanders replied:
Having discussed it, I understand that demo_app.py runs on all 6k bills and this file runs on only 12 bills, and so is faster.

nesanders replied:
Can you please add a doc-level comment explaining this purpose? I also recommend renaming the file, e.g. `demo_app_12bills.py`

nesanders commented on demoapp/demo_app.py:
Thanks for renaming this file!  Can you please edit the README a bit to explain that this is the new working version of the app?

mmomin15 replied:
Also, I forgot to mention in my comment for the commit, but generate_category() function in the original code was using 'gpt-4', and I had changed it to 'gpt-4-1106-preview' to test implementing section text in the prompt. Changing it back to gpt-4.. No other changes were made to the generate_category( ) function.

nesanders replied:
Ok, I am generally in favor of specifying the specific model version (e.g. `gpt-4-1106-preview`) for reproducibility - otherwise, if GPT publishes a new model version, our results may change significantly through no action of our own.

nesanders commented on demoapp/app2.py:
```diff
@@ -15,10 +15,16 @@
 from sidebar import *
 from tagging import *
 
+# This code only implements text summarization, category selection and tagging only for 12 bills
```
Ah, I see you have the comment here. A typicaly doclevel comment in python would be right at the top of the file (line 1) and use a multi-line comment, e.g.

```
"""This is my top level comment.
"""
```

nesanders commented on demoapp/demo_app.py:
```diff
@@ -23,6 +23,10 @@
 from tagging import *
 
 
+# This code only implements text summarization, category selection and tagging for all the bills (no MGL sections data is available for ~1300 bills )
```
Nice, thanks for explaining. Please also move this to line  1 (doclevel comment)

nesanders commented on demoapp/demo_app.py:
```diff
@@ -38,7 +42,7 @@ def get_mgl_sections_file():
         print("May take a few minutes")
         urllib.request.urlretrieve("https://drive.google.com/file/d/1eYMmxW4gLyh7Zxh8BBJTvMOUNtdbK5B3/view?usp=share_link", "demoapp/all_bills_with_mgl.pq")
```
Thanks for hosting this on drive!  Can you add an inline comment explaining how this file was generated (presumably with `extract_mgl_sections.py`?

nesanders replied:
I think this comment has not yet been addressed.

mmomin15 replied:
I added a comment in line 51. Did you mean somewhere else?

nesanders replied:
Oh, I see it now!  It shows up as line 43 in the current version for me. Thanks for explaining!

nesanders commented on demoapp/extract_mgl_sections.py:
```diff
@@ -0,0 +1,319 @@
+import pandas as pd
```
Thanks so much for making this!  Can you add a doclevel comment to explain the purpose?

nesanders replied:
I think this comment has not yet been addressed.

nesanders commented on demoapp/extract_mgl_sections.py:
```diff
@@ -0,0 +1,319 @@
+from requests.packages.urllib3.exceptions import InsecureRequestWarning
+import logging
+
+logging.basicConfig(level=logging.WARNING, format='%(asctime)s - %(levelname)s - %(message)s')
```
Love that you set up logging, great!

nesanders commented on demoapp/extract_mgl_sections.py:
```diff
@@ -0,0 +1,319 @@
+
+logging.basicConfig(level=logging.WARNING, format='%(asctime)s - %(levelname)s - %(message)s')
+
+def csv_to_df_cleaned(bills_csv_file_path, bill_title_col_name, bill_num_col_name, docket_num_col_name, bill_text_col_name):
```
Would be good to add typehints here, e.g.

```suggestion
def csv_to_df_cleaned(bills_csv_file_path: str, bill_title_col_name: str, bill_num_col_name: str, docket_num_col_name: str, bill_text_col_name: str) -> pd.DataFrame:
```

Also, `bills_csv_file_path` could probably be `pathlib.Path | str` type (it's really a path, but pandas will read a string as a path).

nesanders replied:
Not a big deal, but I think this comment has not yet been addressed.

nesanders commented on demoapp/extract_mgl_sections.py:
```diff
@@ -0,0 +1,319 @@
+        logging.error('Error: The file %s was not found', bills_csv_file_path)
+        return None 
+
+def extract_sections(bill_text):
```
Is this the correct output type?

```suggestion
def extract_sections(bill_text: str) -> list[tuple[str, str]]:
```

nesanders replied:
I see you implemented a less detailed type, i.e., `def extract_sections(bill_text: str) -> list: `

Is the full type I suggested above more accurate?

mmomin15 replied:
Yeah, you are right. It is a list of lists. So, your suggestion is more accurate.

nesanders commented on demoapp/extract_mgl_sections.py:
```diff
@@ -0,0 +1,319 @@
+                        if list_with_both not in lists_with_both:
+                            lists_with_both.append(list_with_both)
+                            
+    return lists_with_both
```
Very nicely modularized function, thank you!

nesanders commented on demoapp/extract_mgl_sections.py:
```diff
@@ -0,0 +1,319 @@
+                            
+    return lists_with_both
+
+def apply_extract_sections(bills_df, bill_text_col_name, chap_sec_lists_col_name):
```
```suggestion
def apply_extract_sections(bills_df: pd.DataFrame, bill_text_col_name: str, chap_sec_lists_col_name: str) -> pd.DataFrame:
```

nesanders commented on demoapp/extract_mgl_sections.py:
```diff
@@ -0,0 +1,319 @@
+    the newly created 'Reference List' column.
+
+    Parameters:
+    - bills_df (pd.DataFrame): The DataFrame containing bill information with a 'DocumentText' column.
```
Missing a few params here in the docs, not a huge deal

nesanders commented on demoapp/extract_mgl_sections.py:
```diff
@@ -0,0 +1,319 @@
+    
+    return bills_df
+
+def make_api_call(chapter_section_list):
```
Recommend you rename this to `query_section_text` or similar.

```suggestion
def query_section_text(chapter_section_list: tuple[str, str]) -> str | float:
```

Note that this is really looking for an input tutple (exactly two values, a string and a string) not a list (arbitrary sized container).

Note - the potential `float` output type is for the missing value case (`np.nan`).

nesanders commented on demoapp/extract_mgl_sections.py:
```diff
@@ -0,0 +1,319 @@
+    - The function skips empty or None pairs and ignores pairs with empty or NaN text data.
+    - The formatted text data for each non-empty pair is stored in a list, which is then returned.
+    """
+    # print(df.index)
```
Remove if unused

nesanders replied:
I think this comment has not yet been addressed.

nesanders commented on demoapp/extract_mgl_sections.py:
```diff
@@ -0,0 +1,319 @@
+        pass
+    
+
+def get_mgl_from_api(chapter_section_lists):
```
Recommend renaming to `query_section_text_all_bills` or similar.

```suggestion
def query_section_text_all_bills(chapter_section_lists: list[tuple[str, str]]) -> tuple[list[str], list[tuple[str, str]]]:
```

nesanders commented on demoapp/extract_mgl_sections.py:
```diff
@@ -0,0 +1,319 @@
+    df[word_count_col_name] = df["Combined_MGL"].apply(word_count) # Column name hardcoded for now as it will be used in the demo_app.py
+    return df
+
+def get_df_with_mgl(bills_csv_file_path, bill_title_col_name, bill_num_col_name, docket_num_col_name, bill_text_col_name, chap_sec_lists_col_name, word_count_col_name):
```
I ran out of steam to suggest the last few function signature typehints, but you get the idea :)

nesanders reviewed (commented):
Hi Munira, this is great!  I really like the `extract_mgl_sections.py` module you wrote up. It's really well modularized and will be very useful for us. I left a handful of comments with praise and some suggestions for typehints. This is very close to ready for merge.

mmomin15 committed:
Added typehints for all the functions in extract_mgl_section.py

Renamed app2.py to demo_app_with_12bills.py and added top-level comments

Added top-level comments in demo_app.py and demo_app_with_12bills.py

Added explanation for the demo_app.py in the README

mmomin15 committed:
adding typings, removed print statements

nesanders reviewed (commented):
Great job adding typehinting! There are a few previous comments as yet unaddressed.

mmomin15 committed:
removed print statements, corrected typings,

mmomin15 committed:
Added currently used GPT model versions for generating categories and summaries.
